book/src/rfcs/ffmul.md

```diff
@@ -0,0 +1,816 @@
+ * a) $p_{110} \in [0..2^\ell)$
+ * b) $p_{111} \in [0,3]$
+
+Range check (8) requires two range checks and a decomposition check that is merged in (6).
+ * a) $v_{10} \in [0, 2^{\ell})$
+ * b) $v_{11} \in [0, 7]$
+
+The range checks on $p_0, p_1$ and $p_2$ follow from the range checks on $a,b$ and $q$.
+
+So we have 3.a, 3.b, 4, 7, 8.a, 8.b.
+
+| Range check | Gate type(s)                               | Witness                   | Rows |
+| ----------- | ------------------------------------------ | ------------------------- | ---- |
+| 7           | $(v_0 - 3)(v_0 - 2)(v_0 - 1)v_0$           | $v_0$                     | < 1  |
+| 3.a         | $(p_{111} - 3)(p_{111} - 2)(p_{111} - 1)q$ | $p_{111}$                 | < 1  |
+| 8.b         | degree-8 constraint or plookup             | $v_{11}$                  | 1    |
+| 3.b, 4, 8.a | multi-range-check                          | $p_{10}, p_{110}, v_{10}$ | 4    |
+
+So we have 1 multi-range-check, 1 single-range-check and 2 low-degree range checks. This consumes just over 5 rows.
+
+## Use CRT to constrain $ab - qf - r \equiv 0 \mod n$
+
+We check $ab - qf - r \equiv 0 \mod n$, which is over $\mathbb{F}_n$.
+
+This gives us equality $\mod 2^t \cdot n$ as long as the divisors are coprime.  That is, as long as $\mathsf{gcd}(2^t, n) = 1$.  Since the native modulus $n$ is prime, this is true.
+
+Thus, to perform this check is simple.  We compute
+
+$$
+\begin{aligned}
+a' &= a \mod n \\
+b' &= b \mod n \\
+q' &= q \mod n \\
+f'' &= f \mod n \\
+r' &= r \mod n
+\end{aligned}
+$$
+
+and then constrain
+
+$$
+a' \cdot b' - q' \cdot f'' - r' = 0 \mod n.
+$$
+
+> 9. $a' \cdot b' - q' \cdot f'' - r' = 0 \mod n$
+
+## Range check $q$ so that $q \cdot f + r < 2^t \cdot n$
+
+> The reason this range constraint is needed is that the CRT allows us to verify the desired equality $ab = qf + r$ as integers only when both side of the equation are smaller than $2^t \cdot n$.
+
+In our specific case, $t=264$, $n < 2^{255}$ and $f < 2^{256}$. Then, we can create a constraint for our case that is sufficient to check the above. Concretely, if $q$ was such that $q < 2^{256}$ (even if potentially larger than $f$), we can see that the above claim holds.
+
+If $q < 2^{256}$, and since $f < 2^{256}$ as well, then the main check becomes:
+
+$$
+\begin{aligned}
+  q f + r &< 2^{256} \cdot 2^{256} + 2^{264} = 2^{512} + 2^{264}\\
+  2^t \cdot n &< 2^{264} \cdot 2^{255} \\
+  &\implies \\
+  2^{264} \cdot (2^{248} + 1) &< 2^{264}\cdot 2^{255} \\
+  &\equiv\\
+  (2^{248} + 1) &<  2^{255} \\
+   &\implies \\
+   q f + r &<  2^t \cdot n \quad \text{ if } \quad q < 2^{256}
+\end{aligned}
+$$
+
+Note that the above condition holds by far,
+
+Value $r < f$, so we can check
+
+$$
+\begin{aligned}
+qf + r < qf + f - 1 < qf + f = f(q + 1) &< >2^tn \\
+q &< 2^t\frac{n}{f} - 1.
+\end{aligned}
+$$
+
+However, maybe we should just check
+
+$$
+q < f
+$$
+
+and
+
+$$
+f < 2^t
```

Comment:
This is publicly checkable

Reply:
And in any case, if this was not the case then we wouldn't be able to represent foreign fields with the limbs.

Reply:
agree